File: Data Structures & Algorithms/binary-tree-vertical-order-traversal/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    vector<vector<int>> verticalOrder(TreeNode* root) {
        if(!root) return {};
        unordered_map<int, vector<int>> mp;
        queue<pair<int,TreeNode*>> q;
        q.push({0, root});
        int min_col = 0, max_col = 0;
        while(!q.empty()){
            pair<int,TreeNode*> top = q.front();
            q.pop();
            mp[top.first].push_back(top.second->val);
            min_col = min(min_col, top.first);
            max_col = max(max_col, top.first);
            if(top.second->left){
                q.push({top.first-1, top.second->left});
            }
            if(top.second->right){
                q.push({top.first+1, top.second->right});
            }
        }
        vector<vector<int>> ans;
        for(int i = min_col; i <= max_col; i++){
            ans.push_back(mp[i]);
        }
        return ans;
    }
};

File: Data Structures & Algorithms/binary-tree-diameter/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.ans = 0

        def solve(node):
            if not node:
                return 0
            lh = solve(node.left)
            rh = solve(node.right)
            self.ans = max(self.ans, lh + rh)
            return 1 + max(lh, rh)
        solve(root)
        return self.ans